engine/src/main/java/org/terasology/rendering/dag/nodes/AlphaRejectBlocksNode.java

```diff
@@ -0,0 +1,142 @@
+ */
+package org.terasology.rendering.dag.nodes;
+
+import org.lwjgl.opengl.GL11;
+import org.terasology.assets.ResourceUrn;
+import org.terasology.engine.ComponentSystemManager;
+import org.terasology.math.geom.Vector3f;
+import org.terasology.monitoring.PerformanceMonitor;
+import org.terasology.registry.In;
+import org.terasology.rendering.assets.material.Material;
+import org.terasology.rendering.assets.shader.ShaderProgramFeature;
+import org.terasology.rendering.cameras.Camera;
+import org.terasology.rendering.dag.WireframeCapableNode;
+import org.terasology.rendering.dag.stateChanges.EnableMaterial;
+import org.terasology.rendering.dag.stateChanges.SetViewportToSizeOf;
+import org.terasology.rendering.primitives.ChunkMesh;
+import org.terasology.rendering.world.RenderQueuesHelper;
+import org.terasology.rendering.world.WorldRenderer;
+import org.terasology.world.chunks.ChunkConstants;
+import org.terasology.world.chunks.RenderableChunk;
+
+import static org.terasology.rendering.opengl.DefaultDynamicFBOs.READ_ONLY_GBUFFER;
+
+/**
+ * This node uses alpha-rejection to render semi-transparent blocks (i.e. tree foliage) and
+ * semi-transparent billboards (i.e. plants on the ground).
+ *
+ * Alpha-rejection is the idea that if a fragment has an alpha value lower than some threshold
+ * it gets discarded, leaving the color already stored in the frame buffer untouched.
+ *
+ * This is a less expensive way to render semi-transparent objects compared to alpha-blending.
+ * In alpha-blending the color of a semi-transparent fragment is combined with
+ * the color stored in the frame buffer and the resulting color overwrites the previously stored one.
+ */
+public class AlphaRejectBlocksNode extends WireframeCapableNode {
+
+    private static final ResourceUrn CHUNK_SHADER = new ResourceUrn("engine:prog.chunk");
+
+    @In
+    private ComponentSystemManager componentSystemManager;
+
+    @In
+    private WorldRenderer worldRenderer;
+
+    @In
+    private RenderQueuesHelper renderQueues;
+
+    private Camera playerCamera;
+    private Material chunkShader;
+
+    /**
+     * Initialises the node. -Must- be called once after instantiation.
+     */
+    @Override
+    public void initialise() {
+        super.initialise();
+        playerCamera = worldRenderer.getActiveCamera();
+        addDesiredStateChange(new SetViewportToSizeOf(READ_ONLY_GBUFFER));
+        addDesiredStateChange(new EnableMaterial(CHUNK_SHADER.toString()));
+        chunkShader = getMaterial(CHUNK_SHADER);
+    }
+
+    /**
+     * Renders the world's semi-transparent blocks, i.e. tree foliage and terrain plants.
+     * Does not render fully opaque blocks, i.e. the typical landscape blocks.
+     *
+     * Takes advantage of the two methods
+     *
+     * - WorldRenderer.increaseTrianglesCount(int)
+     * - WorldRenderer.increaseNotReadyChunkCount(int)
+     *
+     * to publish some statistics over its own activity.
+     */
+    @Override
+    public void process() {
```

Comment:
This method in this and other classes now include what was previously WorldRendererImpl.renderChunks().

By moving the code in the separate nodes it can be greatly simplified I feel, making it easier to follow. It also makes it easier to eventually move some lines of it toward the new architecture, taking advantage of State Changes.